lib/kfifo-test.c

```diff
@@ -0,0 +1,219 @@
+
+	kfifo_in(&my_fifo, buffer1, 5);
+	KUNIT_EXPECT_EQ(test, kfifo_len(&my_fifo), 5);
+
+	kfifo_in(&my_fifo, buffer1, 5);
+	KUNIT_EXPECT_EQ(test, kfifo_len(&my_fifo), 10);
+
+	kfifo_reset(&my_fifo);
+	KUNIT_EXPECT_EQ(test, kfifo_len(&my_fifo), 0);
+}
+
+static void kfifo_test_put_should_insert_and_get_should_pop(struct kunit *test)
+{
+	DEFINE_KFIFO(my_fifo, u8, 32);
+
+	u8 out_data = 0;
+	int processed_elements;
+
+	// If the fifo is empty, get returns 0
+	processed_elements = kfifo_get(&my_fifo, &out_data);
+	KUNIT_EXPECT_EQ(test, processed_elements, 0);
+	KUNIT_EXPECT_EQ(test, out_data, 0);
+
+	u8 elements[] = { 3, 5, 11 };
+	for (int i = 0; i < 3; i++) {
+		kfifo_put(&my_fifo, elements[i]);
+	}
+
+	for (int i = 0; i < 3; i++) {
+		processed_elements = kfifo_get(&my_fifo, &out_data);
+		KUNIT_EXPECT_EQ(test, processed_elements, 1);
+		KUNIT_EXPECT_EQ(test, out_data, elements[i]);
+	}
+}
+
+static void kfifo_test_in_should_insert_multiple_elements(struct kunit *test)
+{
+	DEFINE_KFIFO(my_fifo, u8, 32);
+
+	u8 *in_buffer = kunit_kmalloc_array(test, 3, sizeof(u8), GFP_KERNEL);
+	in_buffer[0] = 11;
+	in_buffer[1] = 25;
+	in_buffer[2] = 65;
+
+	kfifo_in(&my_fifo, in_buffer, 3);
+
+	u8 out_data;
+	int processed_elements;
+	processed_elements = kfifo_get(&my_fifo, &out_data);
+	KUNIT_EXPECT_EQ(test, processed_elements, 1);
+	KUNIT_EXPECT_EQ(test, out_data, 11);
+
+	processed_elements = kfifo_get(&my_fifo, &out_data);
+	KUNIT_EXPECT_EQ(test, processed_elements, 1);
+	KUNIT_EXPECT_EQ(test, out_data, 25);
+
+	processed_elements = kfifo_get(&my_fifo, &out_data);
+	KUNIT_EXPECT_EQ(test, processed_elements, 1);
+	KUNIT_EXPECT_EQ(test, out_data, 65);
+}
+
+static void kfifo_test_out_should_pop_multiple_elements(struct kunit *test)
+{
+	DEFINE_KFIFO(my_fifo, u8, 32);
+
+	kfifo_put(&my_fifo, 11);
+	kfifo_put(&my_fifo, 25);
+	kfifo_put(&my_fifo, 65);
+
+	u8 *out_buffer = kunit_kmalloc_array(test, 3, sizeof(u8), GFP_KERNEL);
+	int copied_elements = kfifo_out(&my_fifo, out_buffer, 3);
+	KUNIT_EXPECT_EQ(test, copied_elements, 3);
+
+	KUNIT_EXPECT_EQ(test, out_buffer[0], 11);
+	KUNIT_EXPECT_EQ(test, out_buffer[1], 25);
+	KUNIT_EXPECT_EQ(test, out_buffer[2], 65);
+	KUNIT_EXPECT_TRUE(test, kfifo_is_empty(&my_fifo));
+}
+
+static void kfifo_test_dec_init_should_define_an_empty_fifo(struct kunit *test)
+{
+	DECLARE_KFIFO(my_fifo, u8, 32);
+	INIT_KFIFO(my_fifo);
+
+	// my_fifo is a struct with an inplace buffer
+	KUNIT_EXPECT_FALSE(test, __is_kfifo_ptr(&my_fifo));
+
+	KUNIT_EXPECT_TRUE(test, kfifo_initialized(&my_fifo));
+}
+
+static void kfifo_test_define_should_equal_declare_init(struct kunit *test)
+{
+	// declare a variable my_fifo of type struct kfifo of u8
+	DECLARE_KFIFO(my_fifo1, u8, 32);
+	// initialize the my_fifo variable
+	INIT_KFIFO(my_fifo1);
+
+	// DEFINE_KFIFO declares the variable with the initial value
+	// essentially the same as calling DECLARE_KFIFO and INIT_KFIFO
+	DEFINE_KFIFO(my_fifo2, u8, 32);
+
+	// my_fifo1 and my_fifo2 are equal
```

Comment:
"my_fifo1 and my_fifo2 have the same size"

Reply:
Changed